### ISSUE TITLE
Automatic pagination with .list functions

### ISSUE THREAD

**Issue by joeycozza** [stripe/stripe-node#225]
I was wondering, are there built in options for automatic pagination using the node module?

**Comment by brandur** [comment]
We recently got automatic pagination implemented in Ruby, Java, and Python, but unfortunately Node's is still outstanding. It's certainly something we want to do, but I don't think anyone's currently working on it.

**Comment by brandur** [comment]
I'm going to rename this issue slightly and tag it with "future" so that we have an easy place to track this.

**Comment by TrejGun** [comment]
:+1: for implementation of this feature

**Comment by coffy** [comment]
+1

**Comment by tejasmanohar** [comment]
:+1:

**Comment by brandur** [comment]
Would anyone be willing to propose an interface for this one?

Mostly due to the inherit complexity produced by Node's callbacks, I'm having a bit of trouble coming up with something's that's significantly easier to use than what you'd end up implementing yourself. Ideally we'd want something that could iterate through every result transparently and only make fetches for new pages when necessary, but it seems to me that this would require a callback of the form `(err, obj, next)` or `(err, page, next)` that seems like it would be a major hassle to use.

It's also possible to build a list function that aggregates through _every_ result before trigger a callback, but we've avoided this in other language because it's quite easy to produce enough objects to overwhelm a single process.

I think [coroutines](http://bluebirdjs.com/docs/api/promise.coroutine.html) would present a pretty ideal pagination mechanism, but unfortunately the majority of our Node users seem to be on versions < 0.12.

Would definitely love to hear some ideas, especially from anyone more familiar with idiomatic Node than myself. Thanks!

**Comment by matthewarkin** [comment]
The first thing that comes to mind is something like a stream but streams is node are just ugh.

My other though is like how underscore has a .each so something like 
```
stripe.charges.all( function (err,charge, next/done){
    doSomethingWithCharge();
    done();
})
```
but you'd then need to add like an argument you could send through to done that could stop the pagination which seems a bit hacky and horrible to keep track of...

**Comment by matthewarkin** [comment]
I have used async before to do something similar using queues https://github.com/caolan/async#queue
basically the queue has a worker that processes items in the queue, when the queue is empty add more items to it

**Comment by osuushi** [comment]
Generators/coroutines would be pretty much the perfect interface for this. I'd love to have that, but if most API users are still on older versions of node, I guess it doesn't help much. It would be fantastic to have as an optional add on (e.g. `require('stripe/es6')`) though.

**Comment by brandur** [comment]
@matthewarkin @osuushi Thanks for the feedback!

> Generators/coroutines would be pretty much the perfect interface for this. I'd love to have that, but if most API users are still on older versions of node, I guess it doesn't help much. It would be fantastic to have as an optional add on (e.g. require('stripe/es6')) though.

Nice idea. I'm not sure if it's exactly what you were thinking, but it would be kind of neat if you could have the optional add-on "augment" the existing interface so that if you were on ES6, you could do something like:

``` js
var stripe = require('stripe/next')('key');
```

And just use the package normally except that you'd have access to `stripe.auto_paginate(...)`.

We'd eventually bring the augmented version into mainline and even further along, deprecate and remove the `stripe/next` target.

I'm not hugely involved in the community unfortunately, but do you know if there's any precedent for that kind of pattern?

**Comment by matthewarkin** [comment]
I can't say there is consensus (though I'm not heavily involved in the community to know for certain), node went a `require('stripe').next` method in https://github.com/nodejs/node/pull/5020 

I do like that idea though (of `stripe/next`) and I don't necessarily think it would be too difficult since essentially `stripe/next` could just import stripe and then add those testing apis. 

The name stripe/next sounds like it would be like a v0 where there are no promises to semver though?

**Comment by brandur** [comment]
> The name stripe/next sounds like it would be like a v0 where there are no promises to semver though?

@matthewarkin I was kind of thinking that semver would still exist orthogonally to the `stripe`/`stripe/next` paradigm. If we broke something in either of those includes we'd bump the major version as dictated by semver, etc. Thoughts?

**Comment by matthewarkin** [comment]
Ah I suppose that would work, I suppose it in part depends on what is being planned for next if it's something with a ton of potential changes and stuff it may be weird to have to have a very high semver major version.

I think maybe making next a separate module may be better if there is a general concern about version in since it may be weird for a minor version update of stripe to include a major version change of stripe/next?

**Comment by tejasmanohar** [comment]
-1 for auto-pagination (at least as a priority or default).
+1 for next page helper.

i think even a `next` function in the callback would be a large improvement over the current state of things and is the place to start. coroutines could work, but i'd recommend leaving this to the client (easy to wrap) and keeping the library low-level. until async/await is native and thus, the _de facto_ implementation, it'd be a very opinionated decision. that said, if you choose to take this approach, as long as it's a promise-returning function (bluebird coroutines / co v4), sgtm. hosting a `stripe/next` module within this one seems like a pita from a maintenance perspective and comes packaged with lots of concerns (e.g. semver as we've already seen). why not create a separate module wrapper that accepts a `stripe` client as input and sets a peer dep?

**Comment by floatingLomas** [comment]
I like the idea of just having a `next` function that will grab the next 'page' of results, but how would `next` work for the promise API, though?  I think you could get away with having it return a Collection of Promises and then process that with `each()`, but:

1. There would have to be a way to tell the `list()` function that it should be returning a Collection; and
2. There needs to be a way to 'get out' of the `each()`.

So maybe there does need to be a different `listAll()` that provides for a `next` function in the callback, and returns a Collection of Promises (or at least an Iterable thing) instead of just a single promise, that can be `each()`ed (and some hand-wavey magic way to say "I'm good, no more please").

**Comment by tejasmanohar** [comment]
[this](https://github.com/lucascosta/facebook-js-ads-sdk#edges-and-cursor-based-pagination) seems like a reasonable approach. The response object can contain a `next` function (potentially as a non-enumerable property or we could change the schema of the response callback to be along the lines of `{ data: {}, next: fn, previous: fn }`. It'd definitely be a breaking change though.

**Comment by floatingLomas** [comment]
@tejasmanohar  I like that, and it would totally work - in both a callback and a promise - for general pagination (especially if it also has `hasNext` and `hasPrevious`), and would (should) be non-breaking since it's just new stuff.

That could then be used-in/wrapped-with another, distinct `listAll` (or something) function that gives a single-entry cursor instead of a page.  Both for consistency with all the other language bindings, and for non-breakingness, I don't see how a distinct function can be avoided.*

_* Unless you could pass `cursor: true` (or `iterable: true` or `oneAtATime:true` or some such) as an option to the `list()` function changed its behaviour to make it pass a single object at a time to the callback (vs. one-page-at-a-time), I guess?  This would be opt-in, so (should be) non-breaking.  But it's still a different API from the rest of the bindings..._

**Comment by tejasmanohar** [comment]
If this is the only change that would be better off as a breaking one that the authors of this module have been waiting to release, then +1 on `cursor: true`, but if there are more, just make a new version :)

**Comment by brandur** [comment]
@tejasmanohar @floatingLomas Hello! I just wanted to drop in and say thank-you for the thoughtful discussion here. I'm also a fan of how the linked Facebook ads API module does pagination, and will likely implement something along the lines of that when I get a chance.

I'd lean toward adding a new `list` method. Possibly something like `autoPagingList` to make it look [similar to the naming conventions in other bindings](https://stripe.com/docs/api/ruby#pagination).

**Comment by tejasmanohar** [comment]
Gotcha. So long as auto-paging works in an iterator-type fashion (or calls callbacks incrementally) and doesn't hold everything in memory until the end, sounds good :D

**Comment by ekryski** [comment]
@brandur let me know if you want to collab. Was thinking of digging into this, this week. I'm currently adding a bunch of stuff to https://github.com/feathersjs/feathers-stripe.

**Comment by brandur** [comment]
> @brandur let me know if you want to collab. Was thinking of digging into this, this week. I'm currently adding a bunch of stuff to https://github.com/feathersjs/feathers-stripe and have a bunch of todos in there around handling pagination.

Thanks @ekryski! I want to knock out a few other projects before I get to this one, but I definitely wouldn't mind collaborating here.

**Comment by floatingLomas** [comment]
I had to do some self-pagination (lolz) recently so I dug into this some more - though I still don't have working code to contribute :P - and Bluebird has a bunch of collection-y functional methods like `.map`, `.each`, `.reduce`, etc - so I feel like the "right" way to do this for the promise side is for the `listAll()` method to 'return' a promise-y Iterable, so you can either use one of the collection-y methods to process each result, or you can just do an `.all` or a `.then` and deal with everything.

That said, I'm still not sure how to exit out of this mid-'loop' though in a promise-y way.  RxJs has a `[.doWhile](https://github.com/Reactive-Extensions/RxJS/blob/master/doc/api/core/operators/dowhile.md)` which would totally do this (if we were using Rx streams instead of promises, at least), so maybe we need to ... just add a `.[while](http://blog.victorquinn.com/javascript-promise-while-loop)` loop that allows for an exit...

Anyways, just some more thoughts.

**Comment by brandur** [comment]
> That said, I'm still not sure how to exit out of this mid-'loop' though in a promise-y way. RxJs has a .doWhile which would totally do this (if we were using Rx streams instead of promises, at least), so maybe we need to ... just add a .while loop that allows for an exit...

Oh man, you have to wonder about your language's concurrency model a bit when a while loop becomes a fairly non-trivial implementation ;)

Anyway, your thinking about looks solid. We'll probably need to experiment a bit before landing a definitively good implementation.

**Comment by floatingLomas** [comment]
> Oh man, you have to wonder about your language's concurrency model a bit when an while loop _that includes synchronous array iteration nested inside an unknown number of asynchronous external resource requests_ becomes a fairly non-trivial implementation ;)

FTFY. 😛

I almost feel like `while` should be a Bluebird method, so I feel like we should consider forking/monkeypatching and contributing it back, since `listAll` should ideally be compatible with all of the rest of the list-like methods (`map`, `each`, `filter`, etc) too...

That said, s'more of my thoughts on it:

```
Stripe.thingy.listAll(criteria).while(function onItem(item, next) {
  // next(falseyValue) continues
  // next(truthyValue but !error) breaks and fires .then() with list processed thus far ???
  // next(error) breaks fires .catch().

  // return falseyValue continues
  // return truthyValue (but !errror) breaks and fires .then() ...
  // return error breaks and fires .catch()
 
  // return promiseThatResolves that resolves to falsey continues
  // return promiseThatResolves that resolves to truthey breaks and fires .then() ...
  // return promiseThatRejects breaks and fires .catch()
});
```

...and the callback version could just take `onItem` as an argument (and ignore the bottom three `return` types).

**Comment by nmccready** [comment]
Just some thoughts and walking through the code some.

At the very least there should be an option to return the underlying request (http, https) object. That object is already a stream. This makes handling pagination a breeze, since a stream already gives you the data as you need it. In looking at the [code](https://github.com/stripe/stripe-node/blob/master/lib/StripeResource.js#L217-L226). Line 217 and 226 would need to add returns. Currently everything is tied to the [_responseHandler](https://github.com/stripe/stripe-node/blob/master/lib/StripeResource.js#L110 ) which is wrapped and [promisified](https://github.com/stripe/stripe-node/blob/master/lib/StripeMethod.js#L36-L91). If such a option existed instead of wrapping it just `Promise.resolve(self._request(requestMethod, requestPath, data, opts.auth, options)` where the req is returned because of the options.

I am only suggesting in resolving a stream to not break the api where it seems promised based. However, you could just as easily just return the stream straight up via the option.

Thoughts?

**Comment by msteckyefantis** [comment]
```.js
'use strict';

const stripe = require( 'stripe' )( 'your key' );


module.exports = function listAllAccounts() {

    let accounts = [];

    let values = {

        limit: 100
    };

    return collectAccounts( accounts, values )
        .then( function() {

            return accounts;
        });
}


function collectAccounts( accounts, values ) {

    return stripe.accounts.list( values )
        .then( function( results ) {

            let gottenAccounts = results.data;

            if( gottenAccounts && Array.isArray( gottenAccounts ) ) {

                for( let account of gottenAccounts ) {

                    accounts.push( account );
                }
            }

            if( results.has_more ) {

                let newValues = Object.assign( values, {

                    starting_after: gottenAccounts[ gottenAccounts.length - 1 ].id
                });

                return collectAccounts( accounts, newValues );
            }
        });
}

```

**Comment by billinghamj** [comment]
It definitely makes sense to implement this through generators/iterators, but I would definitely disagree with hiding this behind some kind of special next/ES6 flag.

**Comment by sedubois** [comment]
Are there any news from Stripe on this? If auto-pagination isn't supported by Stripe for Node it could be mentioned in the docs?

It's 2017 and Node has generators/yield/async/await etc out of the box so I'm not quite sure what would prevent providing such auto-pagination (sorry if I missed something while scrolling through the long thread).

Thanks a lot :)

**Comment by joaoreynolds** [comment]
Is this issue being worked on? In the meantime I'll manually iterate through pages, but the docs say this function is coming (however, this is a stale issue)

**Comment by jlomas-stripe** [comment]
I think it's sort of stuck in "the right way is generators" land right now, and the fact that there doesn't appear to be an easy/clean way to do this via Promises (as there's no `Promise.while()` equivalent that would allow exit from the loop).

Maybe we should just implement _something_ callback-based that works and then go from there, because this is a pretty useful feature.

@brandur-stripe - thoughts?  Like, we can't do this 'right' yet (i.e., with generators), and we can't do this 'fully' (i.e., with Promises) at all (it seems), but we could just ... do something that works (with callbacks)...?   ¯\\\_(ツ)_/¯

**Comment by billinghamj** [comment]
Can't this be done with the following new ES syntax?

```js
for await (const result of stripe.customers.listAll())
```

It's a finished proposal (part of ES2018): https://github.com/tc39/proposal-async-iteration

Not widely supported yet, but it isn't going to be that long: http://kangax.github.io/compat-table/es2016plus/#test-Asynchronous_Iterators

Support via Babel is already here: https://babeljs.io/docs/plugins/transform-async-generator-functions/

It shouldn't be hard to expose the underlying bits required for end users to interface with it using the new syntax. Any particular reason not to just go ahead with this now?

**Comment by billinghamj** [comment]
Actually, with the release of Node 10, this is now generally available - with no flags, etc.

**Comment by brandur-stripe** [comment]
@billinghamj Great thought on the async iterators — that looks like the right move.

The only problem is that we need to maintain pretty broad compatibility. Unfortunately, Node V4 is still our most commonly seen runtime by user.

**Comment by billinghamj** [comment]
@brandur-stripe I don't think there's any inherent reason this couldn't be offered with full compatibility. Although the nice special syntax has now been made available, the "under the hood" stuff which makes it work can be put into place without any language support, standard library stuff, etc.

You have to expose the iterator from a field in an object, under `Symbol.asyncIterator`. If the runtime doesn't have Symbol support, then probably can just not export it - as they won't be able to use it anyway.

The async iterator object then exposes a function called `then` which returns a promise.

i.e. I think it'd look (something) like this:

```js
if (typeof Symbol !== 'undefined' && typeof Symbol.asyncIterator !== 'undefined') {
  blah.prototype.listAll = function listAll() {
    function fetchNextItem() {
      // logic for iterating/fetching - would return a promise
    }

    const returnObj = {};

    returnObj[Symbol.asyncIterator] = {
      next: function next() {
        return fetchNextItem();
      },
    };

    return returnObj;
  };
}
```

**Comment by egeste** [comment]
```es6
import stripe from 'stripe'

const stripeClient = stripe(process.env.STRIPE_KEY)

const paginateStripeAPI = (getMore, options = {}, previousList = []) => {
  const latestItem = previousList[previousList.length - 1]
  const starting_after = latestItem ? latestItem.id : undefined

  return getMore({ limit: 3, starting_after, ...options }).then(newList => {
    const fullList = [ ...previousList, ...newList.data ]
    return !newList.has_more ?
      Promise.resolve(fullList) :
      paginateStripeAPI(getMore, options, fullList)
  })
}

// Paginate all invoice items
paginateStripeAPI(options => stripeClient.invoiceItems.list(options))
  .then(invoices => console.log(invoices.length, 'total invoice items'))
  .catch(errv => console.error(err))

// Paginate all charges
paginateStripeAPI(options => stripeClient.charges.list(options))
  .then(invoices => console.log(invoices.length, 'total charges'))
  .catch(errv => console.error(err))

// etc...
```

**Comment by twumm** [comment]
Hi @brandur-stripe !

May I know the timeline for this - if available :) ?
Also, can there be another key similar to `has_more` - **in the invoice list object** -  which tells whether there are previous items?
It will be very helpful in setting state of buttons to active or not.

I will appreciate any feedback on this.

Thanks

**Comment by brandur-stripe** [comment]
@twumm Geeze, just looking at my comments above I realize that apparently I'm a huge liar — sorry about that. I thought I'd finally convinced someone to take a crack at it, but something else came up, and it didn't end up happening.

I'm asked someone better familiar with good JS conventions to look again. If that doesn't work out, remind me again and I'll take a crack at it myself.

@egeste Thanks! This looks like a good start, but I think it'd be better if we could hand back results to users one item or one page at a time as we go. Some of the list endpoints can be somewhat slow and the total number of objects in an account can be enormous, so an ideal solution would be to allow users to handle them as they go.

**Comment by rattrayalex-stripe** [comment]
Hey all! 

I'm finally taking a swing at this – sorry for the delay!

I have a partial implementation towards this vision: 

```js
// To operate on each item, and/or halt partway through:
stripe.customers.list({foo: 'bar'}).autoPagingEach((customer) => {
  if (customer.name === 'Taylor Swift') {
    return false; // Stop iteration.
  }
});

// Or, if you need to perform some asynchronous work on each iteration:
stripe.customers.list({foo: 'bar'}).autoPagingEach((customer, next) => {
  detectTaylorSwift(customer).then((isTaylorSwift) => {
    if (isTaylorSwift) {
      next(false); // Stop iteration.
    } else {
      next(); // Continue, asynchronously.
    }
  });
});

// Or, if your system supports async interators (eg; you are on Node 10), and you don't pass a callback:
for await (const customer of stripe.customers.list({foo: 'bar'}).autoPagingEach()) {
  if (customer.name === 'Taylor Swift') {
    break;
  }
  // Do stuff with `customer`.
}

// To execute code once stopped or complete, you can use a second `onDone` callback:
stripe.customers.list({foo: 'bar'}).autoPagingEach(
  (customer) => {}, // Handle each customer.
  (err, customers) => {} // Handle all customers, or api errors.
);

// or a `.then()`:
stripe.customers.list({foo: 'bar'}).autoPagingEach((customer) => {})
  .then((customers) => {}) // Handle all customers.
  .catch((err) => {}); // Handle api errors.
  
// Which means that if you have `await`, you can do:
const allCustomersArray = await stripe.customers.list({foo: 'bar'}).autoPagingEach();
```

I'm hoping to make more headway on Monday, and maybe put up a WIP PR then. I'm not sure I'll support all the features above, and I haven't quite proved out that everything there is possible (though I think it is). Eg; I'm not sure we'll support _both_ `return false;` and `next(false)`, or if those are the right api's. 

Thoughts?

**Comment by billinghamj** [comment]
Wow, that’s a lot of variations to cover 😬 but looks great!

I could see there being issues with the `return false` version where people forget to return false or return some other falsely value. Due to the significant cost of it continuing to run forever without the developer realising, I wonder whether it’d be better to answer in the affirmative instead - having to specifically `return true` for iteration to continue - rather than the current opposite.

Also does the return of an async iterator conflict with the return of all results if awaited? Given both are shown when no callback is taken in

**Comment by billinghamj** [comment]
Ah also one thing I’d note - probably best to just have `next(true)`, so then unrelated returns don’t have an impact. With setups like Coffeescript or `do` blocks, implicit returns do happen

**Comment by billinghamj** [comment]
And finally, could you perhaps just have `next()` with no args? If you want to continue, call it. If you want to finish, don’t.

**Comment by billinghamj** [comment]
Also I do think that the callback to receive all once it’s done could be problematic.

First, if you’ve stopped part way through, presumably it only gives you the ones it’s seen so far - rather than all of them.

Second, if you’re iterating over a lot, this could quite likely cause an out-of-memory issue. Eg if you had millions of customers and iterated over all of them this way for memory efficiency, you wouldn’t want to find out that the Stripe SDK had been retaining everything in memory unnecessarily.

It also sounds like all of the above will also resolve as promises. I wonder if maybe a single function is trying to do too much? It might be better to just always return an async generator which people can read from manually (without the `for await` syntax sugar) as they need. Then maybe the “all” variant could just be another function.

**Comment by rattrayalex-stripe** [comment]
Thanks for the feedback @billinghamj !

I think you're right that `autoPagingEach` shouldn't fill up memory by default; I'll remove the full list of items from that API, but may add a complementary `autoPagingToArray` helper, either as a method or standalone util. 

> It might be better to just always return an async generator which people can read from manually (without the for await syntax sugar)

Unfortunately, reading from async iterators without `for await` sugar is a bit of a pain, so I'd rather not force everyone into that. 

> could you perhaps just have next() with no args? ...  If you want to finish, don't [call it].

Unfortunately, this would make it impossible to tell when the iteration is "done", which is pretty important (since the promise would just hang if `next()` is never called). 

Related to possible confusion/footgun with `next()`, and limited availability/use of `for await`, I'm still leaning towards not requiring use of the `next` API, but your points re; implicit returns & bifurcation are certainly well-taken.

**Comment by rattrayalex-stripe** [comment]
WIP up at https://github.com/stripe/stripe-node/pull/496

**Comment by rattrayalex-stripe** [comment]
This has been released in [v6.11.0](https://github.com/stripe/stripe-node/releases/tag/v6.11.0)!

Docs are [here](https://github.com/stripe/stripe-node#auto-pagination).

Thanks for your patience, all! Hope you like the result.
